### PR TITLE
Fixing API for `first()` and `count()` functions

### LINKED ISSUE #1177: Fix API of `first` and `count` functions
Currently both functions are defined as `first(iterable, start=0)` and `count(iterable, start=0)`. The following fixes are needed:
- Parameter `start` is not used in these functions -- remove
- When `iterable` is not a `BaseExpr` they do nothing. However, it makes sense to define these functions for an arbitrary python iterable too - as the name of the parameter suggests.
- The `iterable` in `count()` should default to `None`, so that `count()` is valid too. This should return a special `BaseExpr` that computes the number of rows in each group.

In addition, we need to add tests for both of these functions.

### PR DESCRIPTION
**Pull request by oleksiyskononenko**
- make `first()` and `count()` work on `iterables`
- make `count()` work when no column is specified
- make `first()` and `count()` work when there is no `Groupby`
- adding tests

Closes #1177

### CODE REVIEW

File: datatable/expr/reduce_expr.py [hunk 1]
```diff
@@ -18,13 +18,47 @@ def sum(iterable, start=0):
...
+        return _builtin_sum(1 for x in iterable)
 
-def first(iterable, start=0):
+
+def first(iterable):
     if isinstance(iterable, BaseExpr):
         return ReduceExpr("first", iterable)
+    else:
+        for x in iterable:
+            if x:
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply datatable/expr/reduce_expr.py]: Done

File: datatable/expr/reduce_expr.py [hunk 2]
```diff
@@ -18,13 +18,47 @@ def sum(iterable, start=0):
...
 
-def count(iterable, start=0):
+
+def count(iterable=None):
     if isinstance(iterable, BaseExpr):
         return ReduceExpr("count", iterable)
+    elif iterable is None:
+        return CountExpr()
+    else:
+        return _builtin_sum(1 for x in iterable)
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply datatable/expr/reduce_expr.py]: Done

File: datatable/expr/reduce_expr.py [hunk 3]
```diff
@@ -18,13 +18,47 @@ def sum(iterable, start=0):
...
         return ReduceExpr("first", iterable)
+    else:
+        for x in iterable:
+            if x:
+                return x
+              
+              
+class CountExpr(BaseExpr):
+    def __init__(self):
+        super().__init__()
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply datatable/expr/reduce_expr.py]: Done

File: datatable/expr/reduce_expr.py [hunk 4]
```diff
@@ -18,13 +18,47 @@ def sum(iterable, start=0):
...
+
+    def is_reduce_expr(self, ee):
+        return True
+
+    def resolve(self):
+       self._stype = int
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply datatable/expr/reduce_expr.py]: Done

File: datatable/expr/reduce_expr.py [hunk 5]
```diff
@@ -18,13 +18,47 @@ def sum(iterable, start=0):
...
+#         if self._stype is None:
+#             raise ValueError(
+#                 "Cannot compute %s of a variable of type %s"
+#                 % (self._op, expr_stype))
+
+    def evaluate_eager(self, ee):
+        return core.expr_count(ee.groupby)
+
+    def __str__(self):
+        return ""
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply datatable/expr/reduce_expr.py]: Done

File: tests/test_dt_expr.py [hunk 6]
```diff
@@ -282,3 +282,84 @@ def test_div_mod(seed):
         [None if src2[i] == 0 else src1[i] // src2[i] for i in range(n)],
         [None if src2[i] == 0 else src1[i] % src2[i] for i in range(n)]
     ]
+
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply tests/test_dt_expr.py]: Done

**Comment by st-pasha** [comment]
Please merge master into this branch, o/w you won't be able to build it

File: tests/test_dt_expr.py [hunk 7]
```diff
@@ -282,3 +282,84 @@ def test_div_mod(seed):
...
+                              [3, 1, 1, 1, 2, 2, 2, 1]]                        
+
+
+def test_count_2d_array():
+    a = [[9, 8, 2, 3, None, None, 3, 0, 5, 5, 8, None, 1],
+         [0, 1, 0, 5, 3, 8, 1, 0, 2, 5, None, 8, 1]]
+    af = count(a)
+    assert af == 2
+
+
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply tests/test_dt_expr.py]: Done

**Comment by oleksiyskononenko** [comment]
I've merged `master` onto this branch. Working on some of the tests.

File: c/expr/reduceop.cc [hunk 8]
```diff
@@ -41,19 +42,25 @@ static Column* reduce_first(const Column* arg, const Groupby& groupby) {
     return Column::new_data_column(arg->stype(), 0);
   }
   size_t ngrps = groupby.ngroups();
-  // groupby.offsets array has length `ngrps + 1` and contains offsets of the
-  // beginning of each group. We will take this array and reinterpret it as a
-  // RowIndex (taking only the first `ngrps` elements). Applying this rowindex
-  // to the column will produce the vector of first elements in that column.
-  arr32_t indices(ngrps, groupby.offsets_r());
+  arr32_t indices(ngrps);
```
Comment (st-pasha) [inline c/expr/reduceop.cc]: The original constructor `arr32_t indices(ngrps, groupby.offsets_r());` simply uses the pointer as borrowed; however `arr32_t indices(ngrps);` allocates new memory and then you need to copy data into that memory. This is very wasteful.

File: tests/test_reduce.py [hunk 9]
```diff
@@ -39,22 +65,66 @@ def test_count_2d_dt():
...
                               [3, 1, 1, 1, 2, 2, 1, 1],
-                              [3, 1, 1, 1, 2, 2, 2, 1]]                        
+                              [3, 1, 1, 1, 2, 2, 2, 1]]
 
 
-def test_count_2d_array():
-    a = [[9, 8, 2, 3, None, None, 3, 0, 5, 5, 8, None, 1],
-         [0, 1, 0, 5, 3, 8, 1, 0, 2, 5, None, 8, 1]]
+def test_count_array_string():
+    a = [None, "blue", "green", "indico", None, None, "orange", "red", "violet", "yellow", "green", None, "blue"]
```
Comment (st-pasha) [inline tests/test_reduce.py]: Please try not to write lines > 80 chars (unless it actually makes the code more readable).

Comment (oleksiyskononenko) [reply tests/test_reduce.py]: Sure, forgot to adjust my editor for that. Done.

File: c/expr/reduceop.cc (hunk 8 above)
Comment (oleksiyskononenko) [reply c/expr/reduceop.cc]: Pushed implementation that doesn't involve `memcpy`. The segfault problem is quite interesting, for me it appears in release mode only. Will work on it.

**Comment by st-pasha** [comment]
Can you add the following test too:
```
assert df.Frame(A=range(10))[3::3, :][:, first(f.A)].scalar() == 3
```

**Comment by oleksiyskononenko** [comment]
Do you mean `dt.Frame...`?

File: c/expr/py_expr.h [hunk 10]
```diff
@@ -34,8 +34,7 @@ DECLARE_FUNCTION(
 DECLARE_FUNCTION(
   expr_reduceop,
   "expr_reduceop(op, col, groupby)\n\n"
-  "Compute a reduction over the provided column, using the provided Groupby\n"
-  "object.\n",
+  "Compute number of rows in each group, using the provided Groupby object.\n",
```
Comment (st-pasha) [inline c/expr/py_expr.h]: Looks like the documentation of `expr_reduceop` and `expr_count` got mixed up

File: tests/test_dt_expr.py [hunk 11]
```diff
@@ -7,7 +7,7 @@
 import pytest
 import random
 import datatable as dt
-from datatable import f, stype, ltype
+from datatable import f, stype, ltype, first, count
```
Comment (st-pasha) [inline tests/test_dt_expr.py]: looks like you don't need to import first and count here anymore

File: c/expr/py_expr.h (hunk 10 above)
Comment (oleksiyskononenko) [reply c/expr/py_expr.h]: You're right. I've fixed this in one of the commits, not sure how it reverted back.

File: tests/test_dt_expr.py (hunk 11 above)
Comment (oleksiyskononenko) [reply tests/test_dt_expr.py]: Sure, removed.

**Pull request merged**